src/igraph_eulerian.c

```diff
@@ -0,0 +1,614 @@
+    }
+
+    igraph_inclist_destroy(&incl);
+    igraph_vector_destroy(&out_degree);
+    IGRAPH_FINALLY_CLEAN(1);
+
+    return IGRAPH_SUCCESS;
+
+}
+
+
+/* solution adapted from https://www.geeksforgeeks.org/hierholzers-algorithm-directed-graph/ */
+int igraph_i_eulerian_path_directed_implementation(igraph_t *graph, igraph_integer_t *start_node, igraph_vector_t *outgoing_list, igraph_vector_t *res) {
+    igraph_integer_t start = *start_node;
+    igraph_integer_t curr = start;
+    igraph_integer_t next, e_count, curr_e;
+    igraph_inclist_t il;
+    igraph_stack_t path, tracker, edge_tracker, edge_path;
+    igraph_vector_int_t *incedges;
+    igraph_vector_bool_t visited_list;
+    long nc, edge;
+    int j;
+
+    IGRAPH_CHECK(igraph_stack_init(&path, igraph_vcount(graph)));
+    IGRAPH_FINALLY(igraph_stack_destroy, &path);
+
+    IGRAPH_CHECK(igraph_stack_init(&tracker, igraph_vcount(graph)));
+    IGRAPH_FINALLY(igraph_stack_destroy, &tracker);
+
+    IGRAPH_CHECK(igraph_stack_init(&edge_path, igraph_vcount(graph)));
+    IGRAPH_FINALLY(igraph_stack_destroy, &edge_path);
+
+    IGRAPH_CHECK(igraph_stack_init(&edge_tracker, igraph_vcount(graph)));
+    IGRAPH_FINALLY(igraph_stack_destroy, &edge_tracker);
+
+    e_count = igraph_ecount(graph);
+    IGRAPH_CHECK(igraph_vector_bool_init(&visited_list, e_count));
+    IGRAPH_FINALLY(igraph_vector_bool_destroy, &visited_list);
+
+    IGRAPH_CHECK(igraph_stack_push(&tracker, start));
+
+    IGRAPH_CHECK(igraph_inclist_init(graph, &il, IGRAPH_OUT));
+    IGRAPH_FINALLY(igraph_inclist_destroy, &il);
+
+    while (!igraph_stack_empty(&tracker)) {
+        
+        if (VECTOR(*outgoing_list)[curr] != 0) {
+            IGRAPH_CHECK(igraph_stack_push(&tracker, curr));
+            
+            incedges = igraph_inclist_get(&il, curr); /***** INSPECT LATER */
+            nc = igraph_vector_int_size(incedges);
+
+            for (j = 0; j < nc; j++) {
+                edge = (long) VECTOR(*incedges)[j];
+                if (!VECTOR(visited_list)[edge]) {
+                    break;
+                }
+            }
+            
+            next = IGRAPH_TO(graph, edge);
+
+            IGRAPH_CHECK(igraph_stack_push(&edge_tracker, edge));
+
+            /* remove edge here */
+            VECTOR(*outgoing_list)[curr]--;
+            VECTOR(visited_list)[edge] = 1;
+
+            curr = next;
+        } else { /* back track to find remaining circuit */
+            IGRAPH_CHECK(igraph_stack_push(&path, curr));
+            curr = igraph_stack_pop(&tracker);
+            if (!igraph_stack_empty(&edge_tracker)) {
+                curr_e = igraph_stack_pop(&edge_tracker);
+                IGRAPH_CHECK(igraph_stack_push(&edge_path, curr_e));
+            }
+        }
+    }
+
+    while (!igraph_stack_empty(&edge_path)) {
+        IGRAPH_CHECK(igraph_vector_push_back(res, igraph_stack_pop(&edge_path)));           
```

Comment:
You are pushing to the vector `res`, but it has not been cleared. What if the user passes a non-empty vector?

`res` should be [cleared](https://igraph.org/c/doc/igraph-Data-structures.html#igraph_vector_clear) at the start of the function.

We know exactly how many elements the result will have. Therefore, let us [reserve](https://igraph.org/c/doc/igraph-Data-structures.html#igraph_vector_reserve) precisely that amount of storage at the same point where the vector is cleared.